src/frontend/src/catalog/purify.rs

```diff
@@ -53,59 +54,75 @@ pub fn try_purify_table_source_create_sql_ast(
     // Filter out columns that are not defined by users in SQL.
     let defined_columns = columns.iter().filter(|c| c.is_user_defined());
 
-    // If all columns are defined...
-    // - either the schema is fully specified by the user,
-    // - the persisted definition is already purified.
-    // No need to proceed.
+    // If all columns are defined, check if the count matches.
     if !column_defs.is_empty() && wildcard_idx.is_none() {
-        let defined_columns_len = defined_columns.count();
+        let defined_columns_len = defined_columns.clone().count();
         if column_defs.len() != defined_columns_len {
             bail /* unlikely */ !(
                 "column count mismatch: defined {} columns, but {} columns in the definition",
                 defined_columns_len,
                 column_defs.len()
             );
         }
-
-        return Ok(base);
```

Comment:
Previously we short-circuited the procedure once we find all columns are present. Now we should always go through the step for filling the default value.